tracing/span.go

```diff
@@ -201,24 +201,89 @@ func (t *trace) stopAndPublish(ctx context.Context, err error) error {
 	Hooks() []SpanHook
+
+	// AddHooks adds hooks into the Span.
+	//
+	// It is recommended that you only call AddHooks on a Span within an
+	// OnCreate Hook so the Span is set up with all of it's Hooks as a part of
+	// it's creation.
 	AddHooks(hooks ...SpanHook)
-	Start()
+
+	// Stop stops the Span, this is where you can place custom logic for changes
+	// you want to make to the Span once it has ended but before it is
+	// published.
+	//
+	// Stop should not be called directly, instead you should call
+	// tracing.StopSpan() which combines calling span.Stop, span.Publish, and
+	// calling all OnStop Hooks registered to the Span.
+	//
+	// Stop and Publish are separated so you can share a Publish implementation
+	// while customizing Stop.
 	Stop(err error)
+
+	// Publish sends the serialized Span to a back-end that records the Span.
+	//
+	// Publish should not be called directly, instead you should call
+	// span.Stop() which combines stopping the Span, calling span.Publish, and
+	// calling all OnStop Hooks registered to the Span.
+	//
+	// Stop and Publish are separated so you can share a Publish implementation
+	// while customizing Stop.
 	Publish(ctx context.Context, err error) error
+
+	// SetTag sets a binary tag annotation.
+	//
+	// SetTag should not be called directly, instead you should call
+	// tracing.SetTag which combines calling span.SetTag with calling all
+	// OnSetTag Hooks registered to the Span.
 	SetTag(key string, value interface{})
+
+	// AddCounter adds delta to a counter annotation.
+	//
+	// AddCounter should not be called directly, instead you should call
+	// tracing.AddCounter which combines calling span.AddCounter with calling
+	// all OnAddCounter Hooks registered to the Span.
 	AddCounter(key string, delta float64)
+
+	// CreateLocalChild creates a *LocalSpan with given name and component that
+	// is a child of the Span.
+	//
+	// component is an optional, local component name and may be empty string.
+	//
+	// CreateLocalChild should not be called directly, instead use
+	// tracing.CreateLocalChild or tracing.CreateLocalChildForContext which
+	// combines creating the child span with calling any OnCreateChild hooks
+	// registered to the Span.
 	CreateLocalChild(name, component string) *LocalSpan
+
+	// CreateClientChild creates a *ClientSpan with given name that is a child
+	// of the Span.
+	//
+	// A client child-span should be used to make requests to other upstream
+	// servers.
+	//
+	// CreateClientChild should not be called directly, instead use
+	// tracing.CreateClientChild or tracing.CreateClientChildForContext which
+	// combines creating the child span with calling any OnCreateChild hooks
+	// registered to the Span.
 	CreateClientChild(name string) *ClientSpan
+
+	// LogError is a helper method to log an error plus a message.
 	LogError(msg string, err error)
+
+	// SetDebug sets or unsets the debug flag of this Span.
 	SetDebug(v bool)
 }
```

Comment:
I think all the "should not be called directly, use tracing.SameFunction instead" are unnecessary, unless there's a reason the Span itself cannot call those hooks (I don't think so).

Separating `Stop` and `Publish` probably makes sense, but the `Stop` implementation is still supposed to call all the hooks, `tracing. StopSpan` should simply just call `Stop` and `Publish`, nothing else.

Also I don't think `Publish` should take the `err` arg (since it's already taken in `Stop`)?